### PR TITLE
*: enable revive

### PR DESCRIPTION
**Pull request by hawkingrei**
Signed-off-by: Weizhen Wang <wangweizhen@pingcap.com>

### What problem does this PR solve?

Issue Number: ref #35345

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by hawkingrei** [comment]
```revive``` find too many problems. so that it is too hard for me to fix all of them. I will invite the person to fix them.

**Comment by tisonkun** [comment]
/uncc

This patch is incomplete. Also, please describe a bit what a "revive" linter does.

File: util/topsql/topsql.go [hunk 1]
```diff
@@ -167,8 +167,6 @@ func MockHighCPULoad(sql string, sqlPrefixs []string, load int64) bool {
 		if time.Since(start) > 12*time.Millisecond*time.Duration(load) {
 			break
 		}
-		for i := 0; i < 10e5; i++ {
-		}
```
Comment (tangenta) [inline util/topsql/topsql.go:171]: I don't think this can be removed.

Comment (hawkingrei) [reply util/topsql/topsql.go:171]: I think it is an empty loop to wait for something. so why don't they use ```sched_yield```？

Comment (hawkingrei) [reply util/topsql/topsql.go:171]: @crazycs520 PTAL

Comment (hawkingrei) [reply util/topsql/topsql.go]: Sorry, it is my problem. it have been fixed.

File: br/pkg/lightning/log/log.go [hunk 2]
```diff
@@ -47,6 +47,7 @@ type Config struct {
 	FileMaxBackups int `toml:"max-backups" json:"max-backups"`
 }
 
+// Adjust adjusts the log level of the logger.
```
Comment (tangenta) [inline br/pkg/lightning/log/log.go:50]:
```suggestion
// Adjust adjusts some fields in the config to a proper value.
```

File: br/pkg/metautil/metafile.go [hunk 3]
```diff
@@ -55,10 +55,12 @@ const (
 	MetaV2
 )
 
+// CreateMetaFileName creates a meta file name.
```
Comment (tangenta) [inline br/pkg/metautil/metafile.go:58]:
```suggestion
// CreateMetaFileName is the name of meta file.
```

The function name is not so good...

File: br/pkg/storage/azblob.go [hunk 4]
```diff
@@ -389,6 +397,7 @@ func (s *AzureBlobStorage) Create(ctx context.Context, name string) (ExternalFil
 	return uploaderWriter, nil
 }
 
+//Rename is that Remove implements the StorageWriter interface.
```
Comment (tangenta) [inline br/pkg/storage/azblob.go:400]:
```suggestion
// Rename implements the StorageWriter interface.
```

File: br/pkg/storage/local.go [hunk 5]
```diff
@@ -26,13 +26,14 @@ type LocalStorage struct {
 	base string
 }
 
-func (l *LocalStorage) DeleteFile(ctx context.Context, name string) error {
+// DeleteFile deletes the file at `path`.
```
Comment (tangenta) [inline br/pkg/storage/local.go:29]:
```suggestion
// DeleteFile deletes the file.
```

File: br/pkg/storage/azblob.go (hunk 4 above)
Comment (tangenta) [reply br/pkg/storage/azblob.go:400]: Please address this comment.

File: br/pkg/logutil/logging.go [hunk 6]
```diff
@@ -104,6 +104,7 @@ func (t zapStreamBackupTaskInfo) MarshalLogObject(enc zapcore.ObjectEncoder) err
 	return nil
 }
 
+// StreamBackupTaskInfo make the zap fields for a stream backup task info.
```
Comment (lance6716) [inline br/pkg/logutil/logging.go:107]:
```suggestion
// StreamBackupTaskInfo makes the zap fields for a stream backup task info.
```

File: br/pkg/metautil/metafile.go [hunk 7]
```diff
@@ -55,10 +55,12 @@ const (
 	MetaV2
 )
 
+// CreateMetaFileName is the name of meta file.
 func CreateMetaFileName(ts uint64) string {
 	return fmt.Sprintf("%s_%d", MetaFile, ts)
 }
 
+// Encrypt encrypts the content of MetaFile.
```
Comment (lance6716) [inline br/pkg/metautil/metafile.go:63]:
```suggestion
// Encrypt encrypts the content according to CipherInfo.
```

File: br/pkg/metautil/metafile.go [hunk 8]
```diff
@@ -83,6 +85,7 @@ func Encrypt(content []byte, cipher *backuppb.CipherInfo) (encryptedContent, iv
 	}
 }
 
+// Decrypt decrypts the content of MetaFile.
```
Comment (lance6716) [inline br/pkg/metautil/metafile.go:88]:
```suggestion
// Decrypt decrypts the content according to CipherInfo and IV.
```

File: br/pkg/pdutil/pd.go [hunk 9]
```diff
@@ -631,7 +631,7 @@ func (p *PdController) RemoveSchedulers(ctx context.Context) (undo UndoFunc, err
 }
 
 // RemoveSchedulersWithOrigin pause and remove br related schedule configs and return the origin and modified configs
-func (p *PdController) RemoveSchedulersWithOrigin(ctx context.Context) (ClusterConfig, ClusterConfig, error) {
+func (p *PdController) RemoveSchedulersWithOrigin(ctx context.Context) (origin ClusterConfig, modified ClusterConfig, err error) {
```
Comment (lance6716) [inline br/pkg/pdutil/pd.go:634]: can't understand why we need named return values 😵

File: br/pkg/storage/azblob.go [hunk 10]
```diff
@@ -83,6 +84,7 @@ func (options *AzblobBackendOptions) parseFromFlags(flags *pflag.FlagSet) error
 	return nil
 }
 
+// ClientBuilder implements the StorageReader interface.
```
Comment (lance6716) [inline br/pkg/storage/azblob.go:87]:
```suggestion
// ClientBuilder provides common method to build a service client.
```

File: br/pkg/storage/azblob.go [hunk 11]
```diff
@@ -149,23 +151,21 @@ func getAzureServiceClientBuilder(options *backuppb.AzureBlobStorage, opts *Exte
 
 	accountName := options.AccountName
 	if len(accountName) == 0 {
-		if val := os.Getenv("AZURE_STORAGE_ACCOUNT"); len(val) > 0 {
-			accountName = val
-		} else {
+		val := os.Getenv("AZURE_STORAGE_ACCOUNT")
+		if len(val) <= 0 {
```
Comment (lance6716) [inline br/pkg/storage/azblob.go:155]:
```suggestion
		if len(val) = 0 {
```

File: br/pkg/storage/azblob.go [hunk 12]
```diff
@@ -176,15 +176,16 @@ func getAzureServiceClientBuilder(options *backuppb.AzureBlobStorage, opts *Exte
...
+		log.Warn("Failed to get azure token credential but environment variables exist, try to use shared key.", zap.String("tenantId", tenantID), zap.String("clientId", clientID), zap.String("clientSecret", "?"))
 	}
 
 	var sharedKey string
-	if val := os.Getenv("AZURE_STORAGE_KEY"); len(val) > 0 {
-		log.Info("Get azure sharedKey from environment variable $AZURE_STORAGE_KEY")
-		sharedKey = val
-	} else {
+	val := os.Getenv("AZURE_STORAGE_KEY")
+	if len(val) <= 0 {
```
Comment (lance6716) [inline br/pkg/storage/azblob.go:184]:
```suggestion
	if len(val) = 0 {
```

File: br/pkg/storage/gcs.go [hunk 13]
```diff
@@ -258,12 +258,11 @@ func newGCSStorage(ctx context.Context, gcs *backuppb.GCS, opts *ExternalStorage
 				return nil, errors.Annotatef(berrors.ErrStorageInvalidConfig, "%v Or you should provide '--gcs.credentials_file'", err)
 			}
 			if opts.SendCredentials {
-				if len(creds.JSON) > 0 {
-					gcs.CredentialsBlob = string(creds.JSON)
-				} else {
+				if len(creds.JSON) <= 0 {
```
Comment (lance6716) [inline br/pkg/storage/gcs.go:261]:
```suggestion
				if len(creds.JSON) = 0 {
```

**Review by lance6716 — COMMENTED** [review]
rest lgtm (for br/tidb-binlog and some util)

File: meta/autoid/autoid.go [hunk 14]
```diff
@@ -564,7 +564,7 @@ func NewAllocatorsFromTblInfo(store kv.Storage, schemaID int64, tblInfo *model.T
 // but actually we don't care about it, all we need is to calculate the new autoID corresponding to the
 // increment and offset at this time now. To simplify the rule is like (ID - offset) % increment = 0,
 // so the first autoID should be 9, then add increment to it to get 13.
-func (alloc *allocator) Alloc(ctx context.Context, n uint64, increment, offset int64) (int64, int64, error) {
+func (alloc *allocator) Alloc(ctx context.Context, n uint64, increment, offset int64) (min int64, base int64, err error) {
```
Comment (tangenta) [inline meta/autoid/autoid.go:567]:
```suggestion
func (alloc *allocator) Alloc(ctx context.Context, n uint64, increment, offset int64) (min int64, max int64, err error) {
```

File: meta/autoid/autoid.go [hunk 15]
```diff
@@ -728,7 +728,7 @@ func SeekToFirstAutoIDUnSigned(base, increment, offset uint64) uint64 {
 	return nr
 }
 
-func (alloc *allocator) alloc4Signed(ctx context.Context, n uint64, increment, offset int64) (int64, int64, error) {
+func (alloc *allocator) alloc4Signed(ctx context.Context, n uint64, increment, offset int64) (min int64, base int64, err error) {
```
Comment (tangenta) [inline meta/autoid/autoid.go:731]:
```suggestion
func (alloc *allocator) alloc4Signed(ctx context.Context, n uint64, increment, offset int64) (min int64, max int64, err error) {
```

File: meta/autoid/autoid.go [hunk 16]
```diff
@@ -813,12 +813,12 @@ func (alloc *allocator) alloc4Signed(ctx context.Context, n uint64, increment, o
...
 		zap.Int64("table ID", alloc.tbID),
 		zap.Int64("database ID", alloc.dbID))
-	min := alloc.base
+	min = alloc.base
 	alloc.base += n1
 	return min, alloc.base, nil
 }
 
-func (alloc *allocator) alloc4Unsigned(ctx context.Context, n uint64, increment, offset int64) (int64, int64, error) {
+func (alloc *allocator) alloc4Unsigned(ctx context.Context, n uint64, increment, offset int64) (min int64, base int64, err error) {
```
Comment (tangenta) [inline meta/autoid/autoid.go:821]:
```suggestion
func (alloc *allocator) alloc4Unsigned(ctx context.Context, n uint64, increment, offset int64) (min int64, max int64, err error) {
```

**Review by lance6716 — APPROVED** [review]

File: meta/autoid/memid.go [hunk 17]
```diff
@@ -105,7 +105,7 @@ func (alloc *inMemoryAllocator) ForceRebase(requiredBase int64) error {
 	return nil
 }
 
-func (alloc *inMemoryAllocator) alloc4Signed(n uint64, increment, offset int64) (int64, int64, error) {
+func (alloc *inMemoryAllocator) alloc4Signed(n uint64, increment, offset int64) (min int64, base int64, err error) {
```
Comment (tangenta) [inline meta/autoid/memid.go:108]:
```suggestion
func (alloc *inMemoryAllocator) alloc4Signed(n uint64, increment, offset int64) (min int64, max int64, err error) {
```

File: meta/autoid/memid.go [hunk 18]
```diff
@@ -118,12 +118,12 @@ func (alloc *inMemoryAllocator) alloc4Signed(n uint64, increment, offset int64)
...
 	}
 
-	min := alloc.base
+	min = alloc.base
 	alloc.base += n1
 	return min, alloc.base, nil
 }
 
-func (alloc *inMemoryAllocator) alloc4Unsigned(n uint64, increment, offset int64) (int64, int64, error) {
+func (alloc *inMemoryAllocator) alloc4Unsigned(n uint64, increment, offset int64) (min int64, base int64, err error) {
```
Comment (tangenta) [inline meta/autoid/memid.go:126]:
```suggestion
func (alloc *inMemoryAllocator) alloc4Unsigned(n uint64, increment, offset int64) (min int64, max int64, err error) {
```

File: parser/ast/ddl.go [hunk 19]
```diff
@@ -2020,7 +2020,7 @@ const (
 	TableOptionEngine
 	TableOptionCharset
 	TableOptionCollate
-	TableOptionAutoIdCache
+	TableOptionAutoIdCache //nolint:revive
```
Comment (tangenta) [inline parser/ast/ddl.go:2023]: Why is this marked as "nolint"?

File: parser/model/ddl.go [hunk 20]
```diff
@@ -638,10 +642,12 @@ func (job *Job) IsRunning() bool {
 	return job.State == JobStateRunning
 }
 
+// IsQueueing returns whether job is queuing or not.
 func (job *Job) IsQueueing() bool {
 	return job.State == JobStateQueueing
 }
 
+// NotStarted returns whether job is not started or not.
```
Comment (tangenta) [inline parser/model/ddl.go:650]:
```suggestion
// NotStarted returns true if the job is never run by a worker.
```

File: parser/model/ddl.go [hunk 21]
```diff
@@ -270,13 +271,15 @@ type MultiSchemaInfo struct {
...
 
+// NewMultiSchemaInfo new a MultiSchemaInfo.
 func NewMultiSchemaInfo() *MultiSchemaInfo {
 	return &MultiSchemaInfo{
 		SubJobs:    nil,
 		Revertible: true,
 	}
 }
 
+// SubJob is a job that will be done in a sub-process.
```
Comment (tangenta) [inline parser/model/ddl.go:282]:
```suggestion
// SubJob is a representation of one DDL schema change. A Job may contain zero(when multi-schema change is not applicable) or more SubJobs.
```

File: parser/terror/terror.go [hunk 22]
```diff
@@ -250,8 +251,10 @@ func getMySQLErrorCode(e *Error) uint16 {
 
 var (
 	// ErrClassToMySQLCodes is the map of ErrClass to code-set.
-	ErrClassToMySQLCodes  = make(map[ErrClass]map[ErrCode]struct{})
-	ErrCritical           = ClassGlobal.NewStdErr(CodeExecResultIsEmpty, mysql.Message("critical error %v", nil))
+	ErrClassToMySQLCodes = make(map[ErrClass]map[ErrCode]struct{})
+	// ErrCritical is the critical error class.
+	ErrCritical = ClassGlobal.NewStdErr(CodeExecResultIsEmpty, mysql.Message("critical error %v", nil))
+	// ErrResultUndetermined is the result undetermined error class.
```
Comment (tangenta) [inline parser/terror/terror.go:257]:
```suggestion
	// ErrResultUndetermined is the error when execution result is unknown.
```

File: tidb-binlog/pump_client/client.go [hunk 23]
```diff
@@ -337,7 +337,7 @@ func (c *PumpsClient) WriteBinlog(binlog *pb.Binlog) error {
 
 // Return directly for non p-binlog.
 // Try every online pump for p-binlog.
-func (c *PumpsClient) backoffWriteBinlog(req *pb.WriteBinlogReq, binlogType pb.BinlogType, startTS int64) (pump *PumpStatus, err error) {
+func (c *PumpsClient) backoffWriteBinlog(req *pb.WriteBinlogReq, binlogType pb.BinlogType, _ int64) (pump *PumpStatus, err error) {
```
Comment (tangenta) [inline tidb-binlog/pump_client/client.go:340]: Is it a bug?

File: util/collate/pinyin_tidb_as_cs.go [hunk 24]
```diff
@@ -18,22 +18,22 @@ package collate
 type zhPinyinTiDBASCSCollator struct {
 }
 
-// Collator interface, no implements now.
-func (py *zhPinyinTiDBASCSCollator) Compare(a, b string) int {
+// Compare no implements now.
```
Comment (tangenta) [inline util/collate/pinyin_tidb_as_cs.go:21]:
```suggestion
// Compare is not implemented.
```

File: util/collate/pinyin_tidb_as_cs.go [hunk 25]
```diff
@@ -18,22 +18,22 @@ package collate
...
+func (*zhPinyinTiDBASCSCollator) Compare(_, _ string) int {
 	panic("implement me")
 }
 
-// Collator interface, no implements now.
-func (py *zhPinyinTiDBASCSCollator) Key(str string) []byte {
+// Key no implements now.
```
Comment (tangenta) [inline util/collate/pinyin_tidb_as_cs.go:26]:
```suggestion
// Key is not implemented.
```

File: util/collate/pinyin_tidb_as_cs.go [hunk 26]
```diff
@@ -18,22 +18,22 @@ package collate
...
+func (*zhPinyinTiDBASCSCollator) Key(_ string) []byte {
 	panic("implement me")
 }
 
-// Collator interface, no implements now.
-func (py *zhPinyinTiDBASCSCollator) KeyWithoutTrimRightSpace(str string) []byte {
+// KeyWithoutTrimRightSpace no implements now.
```
Comment (tangenta) [inline util/collate/pinyin_tidb_as_cs.go:31]:
```suggestion
// KeyWithoutTrimRightSpace is not implemented.
```

File: util/collate/pinyin_tidb_as_cs.go [hunk 27]
```diff
@@ -18,22 +18,22 @@ package collate
...
+func (*zhPinyinTiDBASCSCollator) KeyWithoutTrimRightSpace(_ string) []byte {
 	panic("implement me")
 }
 
-// Collator interface, no implements now.
-func (py *zhPinyinTiDBASCSCollator) Pattern() WildcardPattern {
+// Pattern no implements now.
```
Comment (tangenta) [inline util/collate/pinyin_tidb_as_cs.go:36]:
```suggestion
// Pattern is not implemented.
```

File: parser/ast/ddl.go (hunk 19 above)
Comment (hawkingrei) [reply parser/ast/ddl.go:2023]: because ```Id``` should be ```ID```.

File: tidb-binlog/pump_client/client.go (hunk 23 above)
Comment (hawkingrei) [reply tidb-binlog/pump_client/client.go:340]: Maybe, But is it useful?

File: br/pkg/pdutil/pd.go (hunk 9 above)
Comment (disksing) [inline br/pkg/pdutil/pd.go:634]: the variable names are never used.

File: util/table-filter/parser.go [hunk 28]
```diff
@@ -275,13 +275,12 @@ parseLoop:
 			i = end
 
 		default:
-			if c == '$' || c == '_' || isASCIIAlphanumeric(c) || c >= 0x80 {
-				literalStringBuilder.WriteByte(c)
-				wildcardPatternBuilder.WriteByte(c)
-				i++
-			} else {
+			if !(c == '$' || c == '_' || isASCIIAlphanumeric(c) || c >= 0x80) {
```
Comment (disksing) [inline util/table-filter/parser.go:278]: maybe `c != '$' && c != '_' && ...` would be more clear.

**Comment by disksing** [comment]
I think some rules can be relaxed. This PR is too huge to review...

**Comment by hawkingrei** [comment]
@tangenta PTAL

**Comment by lance6716** [comment]
hope it can be merged quickly. Otherwise @hawkingrei will be busy fixing new commits in master :joy:

File: parser/charset/charset.go [hunk 29]
```diff
@@ -52,7 +54,7 @@ var collationsIDMap = make(map[int]*Collation)
 var collationsNameMap = make(map[string]*Collation)
 var supportedCollations = make([]*Collation, 0, len(supportedCollationNames))
 
-// CharacterSetInfos: All the supported charsets should be in the following table.
+// CharacterSetInfos All the supported charsets should be in the following table.
```
Comment (tangenta) [inline parser/charset/charset.go:57]:
```suggestion
// CharacterSetInfos contains all the supported charsets.
```

File: parser/charset/charset.go [hunk 30]
```diff
@@ -158,7 +160,7 @@ func GetCharsetInfo(cs string) (*Charset, error) {
 }
 
 // GetCharsetInfoByID returns charset and collation for id as cs_number.
-func GetCharsetInfoByID(coID int) (string, string, error) {
+func GetCharsetInfoByID(coID int) (defaultCharset string, defaultCollationName string, err error) {
```
Comment (tangenta) [inline parser/charset/charset.go:163]:
```suggestion
func GetCharsetInfoByID(coID int) (charsetStr string, collateStr string, err error) {
```

File: parser/format/format.go [hunk 31]
```diff
@@ -330,6 +336,7 @@ func (ctx *RestoreCtx) WriteKeyWord(keyWord string) {
 	fmt.Fprint(ctx.In, keyWord)
 }
 
+// WriteWithSpecialComments writes the `str` into writer.
```
Comment (tangenta) [inline parser/format/format.go:339]:
```suggestion
// WriteWithSpecialComments writes a string with a special comment wrapped.
```

File: parser/yy_parser.go [hunk 32]
```diff
@@ -70,12 +70,17 @@ func TrimComment(txt string) string {
 	return specCodeEnd.ReplaceAllString(txt, "")
 }
 
+//revive:disable:exported
+
+//ParserConfig is the parser config.
```
Comment (tangenta) [inline parser/yy_parser.go:75]:
```suggestion
// ParserConfig is the parser config.
```

File: types/time.go [hunk 33]
```diff
@@ -1695,16 +1693,18 @@ func matchDuration(str string, fsp int) (Duration, bool, error) {
...
-	} else if hms, remain, err := matchHHMMSSDelimited(rest, true); err == nil {
-		rest, hhmmss = remain, hms
-	} else if hms, remain, err := matchHHMMSSCompact(rest); err == nil {
+	} else if hms, remain, err = matchHHMMSSDelimited(rest, true); err == nil {
 		rest, hhmmss = remain, hms
-	} else {
+	} else if hms, remain, err = matchHHMMSSCompact(rest); err != nil {
 		return ZeroDuration, true, ErrTruncatedWrongVal.GenWithStackByArgs("time", str)
+	} else {
+		rest, hhmmss = remain, hms
```
Comment (tangenta) [inline types/time.go:1707]: The original logic is much clear.

File: util/ranger/points.go [hunk 34]
```diff
@@ -220,7 +220,7 @@ func (r *builder) buildFromConstant(expr *expression.Constant) []*point {
 	return getFullRange()
 }
 
-func (r *builder) buildFromColumn(expr *expression.Column) []*point {
+func (*builder) buildFromColumn(_ *expression.Column) []*point {
```
Comment (tangenta) [inline util/ranger/points.go:223]: What? `buildFromColumn` doesn't use column?

File: types/json/binary_functions.go [hunk 35]
```diff
@@ -379,19 +379,19 @@ func (bj BinaryJSON) Modify(pathExprList []PathExpression, values []BinaryJSON,
 // ArrayInsert insert a BinaryJSON into the given array cell.
 // All path expressions cannot contain * or ** wildcard.
 // If any error occurs, the input won't be changed.
-func (bj BinaryJSON) ArrayInsert(pathExpr PathExpression, value BinaryJSON) (res BinaryJSON, err error) {
+func (bj *BinaryJSON) ArrayInsert(pathExpr PathExpression, value BinaryJSON) (res BinaryJSON, err error) {
```
Comment (tangenta) [inline types/json/binary_functions.go:382]: Why is it changed to a pointer?

File: util/expensivequery/memory_usage_alarm.go [hunk 36]
```diff
@@ -229,24 +229,26 @@ func (record *memoryUsageAlarm) recordProfile() {
...
-		defer func() {
-			err := f.Close()
+		func() {
+			fileName := filepath.Join(record.tmpDir, item.name+record.lastCheckTime.Format(time.RFC3339))
+			record.lastProfileFileName[i] = append(record.lastProfileFileName[i], fileName)
+			f, err := os.Create(fileName)
 			if err != nil {
-				logutil.BgLogger().Error(fmt.Sprintf("close %v profile file fail", item.name), zap.Error(err))
+				logutil.BgLogger().Error(fmt.Sprintf("create %v profile file fail", item.name), zap.Error(err))
+				return
```
Comment (tangenta) [inline util/expensivequery/memory_usage_alarm.go:238]: The `return` semantic is changed here.

File: util/table-filter/parser.go [hunk 37]
```diff
@@ -275,13 +275,12 @@ parseLoop:
...
-				literalStringBuilder.WriteByte(c)
-				wildcardPatternBuilder.WriteByte(c)
-				i++
-			} else {
+			if c != '$' && c != '_' && !isASCIIAlphanumeric(c) && c < 0x80 {
 				return nil, "", p.errorf("unexpected special character '%c'", c)
 			}
+			literalStringBuilder.WriteByte(c)
+			wildcardPatternBuilder.WriteByte(c)
+			i++
```
Comment (tangenta) [inline util/table-filter/parser.go:283]: The original logic is much clear..

File: tidb-binlog/pump_client/selector.go [hunk 38]
```diff
@@ -132,7 +132,7 @@ func (r *RangeSelector) SetPumps(pumps []*PumpStatus) {
 }
 
 // Select implement PumpSelector.Select.
-func (r *RangeSelector) Select(binlog *pb.Binlog, retryTime int) *PumpStatus {
+func (r *RangeSelector) Select(binlog *pb.Binlog, _ int) *PumpStatus {
```
Comment (tangenta) [inline tidb-binlog/pump_client/selector.go:135]: How about removing the unused parameter?

Comment (hawkingrei) [reply tidb-binlog/pump_client/selector.go]: removed

File: util/table-filter/parser.go (hunk 37 above)
Comment (hawkingrei) [reply util/table-filter/parser.go]: revert

File: util/expensivequery/memory_usage_alarm.go (hunk 36 above)
Comment (hawkingrei) [reply util/expensivequery/memory_usage_alarm.go:238]: revert

File: tidb-binlog/pump_client/selector.go (hunk 38 above)
Comment (hawkingrei) [reply tidb-binlog/pump_client/selector.go:135]: It cannot be removed. it is to implement this interface and this interface is important.

File: types/json/binary_functions.go (hunk 35 above)
Comment (hawkingrei) [reply types/json/binary_functions.go]: revert

File: util/ranger/points.go (hunk 34 above)
Comment (hawkingrei) [reply util/ranger/points.go:223]: It is very interesting. I don't know why.

File: types/time.go (hunk 33 above)
Comment (hawkingrei) [reply types/time.go:1707]: revert

File: util/ranger/points.go (hunk 34 above)
Comment (hawkingrei) [reply util/ranger/points.go:223]: remove the useless

**Review by tangenta — APPROVED** [review]

**Pull request merged**